File: generators/cs/MD5.cs
using System;
using System.Collections.Generic;

namespace HashFunctions
{
    public class MD5
    {
        UInt32 mask = 0xFFFFFFFF;
        int width = 32;

        List<UInt32> words = new List<UInt32> { 0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476 };

        private UInt32 F(UInt32 x, UInt32 y, UInt32 z)
        {
            return (x & y) | (~x & z);
        }

        private UInt32 G(UInt32 x, UInt32 y, UInt32 z)
        {
            return (z & x) | (~z & y);
        }

        private UInt32 H(UInt32 x, UInt32 y, UInt32 z)
        {
            return (x ^ y ^ z);
        }
        private UInt32 I(UInt32 x, UInt32 y, UInt32 z)
        {
            return y ^ (x | ~z);
        }

        private UInt32 leftRotate(UInt64 val, int n)
        {
            UInt32 lbits = (UInt32)((val << n) & mask);
            UInt32 rbits = (UInt32)(val >> (width - n));
            return lbits | rbits;
        }

        public void init(string input)
        {
            List<UInt32> message = new List<UInt32>();

            for (int i = 0; i < input.Length; i++)
            {
                message.Add(Convert.ToUInt32(input[i]));
            }

            UInt64 length = (UInt64)(message.Count * 8);

            message.Add(0x80);

            while ((message.Count * 8 + 64) % 512 != 0)
            {
                message.Add(0x00);
            }

            // Add padding
            for (int i = 0; i <= 56; i += 8)
                message.Add((uint)(length >> i));

            List<List<UInt32>> message_chunks = new List<List<UInt32>>();

            for (int i = 0; i < message.Count * 8 / 512; i++)
            {
                List<UInt32> vtemp = new List<UInt32>();

                for (int j = 0; j < 64; j++)
                {
                    vtemp.Add(message[(i * 64) + j]);
                }

                message_chunks.Add(vtemp);
            }

            process(message_chunks);
        }

        public string Digest(string message)
        {
            init(message);

            string s = "";
            foreach (UInt32 word in words)
            {
                byte[] bytes = BitConverter.GetBytes(word);
                string retval = "";
                foreach (byte b in bytes)
                    retval += b.ToString("x2");
                s += retval;
            }

            return s;
        }

        private void process(List<List<UInt32>> chunks)
        {
            foreach (List<UInt32> chunk in chunks)
            {
                List<UInt32> X = new List<UInt32>();
                List<UInt32> h = new List<UInt32>(words);

                //prepare constants
                List<UInt32> constants = new List<UInt32>();
                for (int i = 0; i < 64; i++)
                {
                    UInt64 c = ((ulong)(Math.Abs((double)Math.Sin(i + 1)) * 4294967296));
                    c = c & mask;
                    constants.Add((uint)c);
                }

                //bytes to int
                for (int j = 0; j < 64; j = j + 4)
                {
                    UInt32 temp = 0x00;

                    temp = (UInt32)chunk[j + 3] << 24;
                    temp |= (UInt32)chunk[j + 2] << 16;
                    temp |= (UInt32)chunk[j + 1] << 8;
                    temp |= (UInt32)chunk[j + 0];

                    X.Add(temp);
                }

                List<UInt32> Xi = new List<UInt32> { 7, 12, 17, 22 };
                // round 1
                for (int n = 0; n < 16; n++)
                {
                    //modulo w C++ działa inaczej na negatywnych intach niż python dlatego potrzebne są takie komplikacje
                    int a = (4 + ((-n + 0) % 4)) % 4;
                    int b = (4 + ((-n + 1) % 4)) % 4;
                    int c = (4 + ((-n + 2) % 4)) % 4;
                    int d = (4 + ((-n + 3) % 4)) % 4;

                    int K = n;
                    int S = (int)Xi[n % 4];
                    int T = n;


                    //wartości tej zmiennej wykraczają poza 32 bity, dlatego trzeba je rzutować na ullong
                    UInt64 to_rotate = h[a];
                    to_rotate += X[K];
                    to_rotate += F(h[b], h[c], h[d]);
                    to_rotate += constants[T];

                    h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
                }

                Xi = new List<UInt32> { 5, 9, 14, 20 };
                // round 2
                for (int n = 0; n < 16; n++)
                {
                    int a = (4 + ((-n + 0) % 4)) % 4;
                    int b = (4 + ((-n + 1) % 4)) % 4;
                    int c = (4 + ((-n + 2) % 4)) % 4;
                    int d = (4 + ((-n + 3) % 4)) % 4;

                    double te = (5 * n + 1) % 16;
                    int K = (int)Math.Floor(te);
                    int S = (int)Xi[n % 4];
                    int T = n + 16;

                    UInt64 to_rotate = h[a];
                    to_rotate += X[K];
                    to_rotate += G(h[b], h[c], h[d]);
                    to_rotate += constants[T];

                    h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
                }

                Xi = new List<UInt32> { 4, 11, 16, 23 };
                // round 3
                for (int n = 0; n < 16; n++)
                {
                    int a = (4 + ((-n + 0) % 4)) % 4;
                    int b = (4 + ((-n + 1) % 4)) % 4;
                    int c = (4 + ((-n + 2) % 4)) % 4;
                    int d = (4 + ((-n + 3) % 4)) % 4;

                    int K = (3 * n + 5) % 16;
                    int S = (int)Xi[n % 4];
                    int T = n + 32;

                    UInt64 to_rotate = h[a];
                    to_rotate += X[K];
                    to_rotate += H(h[b], h[c], h[d]);
                    to_rotate += constants[T];

                    h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
                }

                Xi = new List<UInt32> { 6, 10, 15, 21 };
                // round 4
                for (int n = 0; n < 16; n++)
                {
                    int a = (4 + ((-n + 0) % 4)) % 4;
                    int b = (4 + ((-n + 1) % 4)) % 4;
                    int c = (4 + ((-n + 2) % 4)) % 4;
                    int d = (4 + ((-n + 3) % 4)) % 4;

                    int K = (7 * n) % 16;
                    int S = (int)Xi[n % 4];
                    int T = n + 48;

                    UInt64 to_rotate = h[a];
                    to_rotate += X[K];
                    to_rotate += I(h[b], h[c], h[d]);
                    to_rotate += constants[T];

                    h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
                }

                words[0] = (words[0] + h[0]) & mask;
                words[1] = (words[1] + h[1]) & mask;
                words[2] = (words[2] + h[2]) & mask;
                words[3] = (words[3] + h[3]) & mask;
            }
        }
    }
}

File: generators/cpp/md5.cpp
#include "md5.h"
using namespace std;

ulong MD5::F(ulong x, ulong y, ulong z) {
    return (x & y) | (~x & z);
}

ulong MD5::G(ulong x, ulong y, ulong z) {
    return (z & x) | (~z & y);
}

ulong MD5::H(ulong x, ulong y, ulong z) {
    return x ^ y ^ z;
}

ulong MD5::I(ulong x, ulong y, ulong z) {
    return y ^ (x | ~z);
}

ulong MD5::leftRotate(unsigned long long val, int n) {
    ulong lbits = (val << n) & mask;
    ulong rbits = val >> (width - n);
    return lbits | rbits;
}


vector<ulong> MD5::resizeBlock(vector<ulong> in) {
    vector<ulong> output(16);

    for (int i = 0; i < 64; i = i + 4) {
        bitset<32> temp(0);

        temp = (ulong)in[i] << 24;
        temp |= (ulong)in[i + 1] << 16;
        temp |= (ulong)in[i + 2] << 8;
        temp |= (ulong)in[i + 3];

        output[i / 4] = temp.to_ulong();
    }

    return output;
}

MD5::MD5(string input) {
    vector<ulong> message;

    for (unsigned i = 0; i < input.size(); ++i) {
        bitset<8> b(input.c_str()[i]);
        message.push_back(b.to_ulong());
    }

    //uint32_t length = message.size() * 8;
    ull ll = static_cast<ull>(message.size() * 8);

    //dodanie bitu 128 na koniec wiadomości
    message.push_back(0x80);

    //uzupełnie wiadomości do 448 bitów
    while ((message.size() * 8 + 64) % 512 != 0) {
        message.push_back(0x00);
    }

    //dodanie padingu do 512 bitów
    for (int i = 0; i <= 56; i += 8) {
        message.push_back(ll >> i);
    }

    //podział wiadomości na bloki 512 bitów
    vector<vector<ulong>> message_chunks;
    for (unsigned i = 0; i < message.size() * 8 / 512; i++) {
        vector<ulong> vtemp;

        for (int j = 0; j < 64; j++) {
            vtemp.push_back(message[(i * 64) + j]);
        }

        message_chunks.push_back(vtemp);
    }

    process(message_chunks);
}

vector<unsigned char> MD5::getHex(ulong in) {
    string temp;
    vector<unsigned char> b = { 0, 0, 0, 0 };
    b[3] = (int)((words[in] >> 24) & 0xFF);
    b[2] = (int)((words[in] >> 16) & 0xFF);
    b[1] = (int)((words[in] >> 8) & 0XFF);
    b[0] = (int)((words[in] & 0XFF));

    return b;
}

string MD5::hexdigest() {
    string res = "";

    std::stringstream ss;
    ss << std::hex;
    for (int wi = 0; wi < 4; wi++) {
        for (int i(0); i < 4; i++)
            ss << std::setw(2) << std::setfill('0') << (int)getHex(wi)[i];
    }
    ss >> res;
    return res;
}

void MD5::process(vector<vector<ulong>> chunks) {
    for (vector<ulong> chunk : chunks) {
        vector<ulong> X;
        vector<ulong> h = words;

        //prepare constants
        vector<ulong> constants;
        for (int i = 0; i < 64; i++) {
            ull c = (abs( (double)sin(i + 1)) * 4294967296);
            c = c & mask;
            constants.push_back(c);
            //cout << constants[i] << "  ";
        }

        for (int i = 0; i < chunk.size(); i += 4) {
            unsigned char bytes[4]{ chunk[i], chunk[i + 1], chunk[i + 2], chunk[i + 3] };
            unsigned int value;
            std::memcpy(&value, bytes, sizeof(int));
            X.push_back(value);
        }

        vector<ulong> Xi = { 7, 12, 17, 22 };
        // round 1
        for (int n = 0; n < 16; n++) {
            //modulo w C++ działa inaczej na negatywnych intach niż python dlatego potrzebne są takie komplikacje
            int a = (4 + ((-n + 0) % 4)) % 4;
            int b = (4 + ((-n + 1) % 4)) % 4;;
            int c = (4 + ((-n + 2) % 4)) % 4;;
            int d = (4 + ((-n + 3) % 4)) % 4;;

            int K = n;
            int S = Xi[n % 4];
            int T = n;

            //wartości tej zmiennej wykraczają poza 32 bity, dlatego trzeba je rzutować na ullong
            ull to_rotate = h[a];
            to_rotate += X[K];
            to_rotate += F(h[b], h[c], h[d]);
            to_rotate += constants[T];

            h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
        }

        //round 2
        Xi = { 5, 9, 14, 20 };
        for (int n = 0; n < 16; n++) {
            int a = (4 + ((-n + 0) % 4)) % 4;
            int b = (4 + ((-n + 1) % 4)) % 4;;
            int c = (4 + ((-n + 2) % 4)) % 4;;
            int d = (4 + ((-n + 3) % 4)) % 4;;

            int K = (5 * n + 1) % 16;
            int S = Xi[n % 4];
            int T = n + 16;

            ull to_rotate = h[a];
            to_rotate += X[K];
            to_rotate += G(h[b], h[c], h[d]);
            to_rotate += constants[T];

            h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
        }

        //round 3
        Xi = { 4, 11, 16, 23 };
        for (int n = 0; n < 16; n++) {
            int a = (4 + ((-n + 0) % 4)) % 4;
            int b = (4 + ((-n + 1) % 4)) % 4;;
            int c = (4 + ((-n + 2) % 4)) % 4;;
            int d = (4 + ((-n + 3) % 4)) % 4;;

            int K = (3 * n + 5) % 16;
            int S = Xi[n % 4];
            int T = n + 32;

            ull to_rotate = h[a];
            to_rotate += X[K];
            to_rotate += H(h[b], h[c], h[d]);
            to_rotate += constants[T];

            h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
        }

        //round 4
        Xi = { 6, 10, 15, 21 };
        for (int n = 0; n < 16; n++) {
            int a = (4 + ((-n + 0) % 4)) % 4;
            int b = (4 + ((-n + 1) % 4)) % 4;;
            int c = (4 + ((-n + 2) % 4)) % 4;;
            int d = (4 + ((-n + 3) % 4)) % 4;;

            int K = (7 * n) % 16;
            int S = Xi[n % 4];
            int T = n + 48;

            ull to_rotate = h[a];
            to_rotate += X[K];
            to_rotate += I(h[b], h[c], h[d]);
            to_rotate += constants[T];

            h[a] = (h[b] + leftRotate((to_rotate & mask), S)) & mask;
        }

        words[0] = (words[0] + h[0]) & mask;
        words[1] = (words[1] + h[1]) & mask;
        words[2] = (words[2] + h[2]) & mask;
        words[3] = (words[3] + h[3]) & mask;
    }
}

/*
int main() {
    MD5 md5("hello");
    string s = md5.hexdigest();
    cout << s << "\n";

    return 0;
}
*/